text/0477-policy-validation.md

```diff
@@ -0,0 +1,739 @@
+
+  Recommendation: Missing value for key `SSEAlgorithm` - must specify `aws:kms`
+  How to fix:
+    > Using override `app.findChild('my-bucket').addPropertyOverride('SSEAlgorithm', 'aws:kms');`
+    > Add to construct properties for `cdk-app/MyStack/Bucket`
+      `encryption: BucketEncryption.KMS`
+
+  
+Validation failed. See above reports for details
+```
+
+By default, the report will be printed in a human readable format. If you want a
+report in JSON format, enable it using the `@aws-cdk/core:validationReportJson`
+via the CLI or passing it directly to the application:
+
+```ts
+const app = new App({ 
+  context: { '@aws-cdk/core:validationReportJson': true }, 
+});
+```
+
+### Plugins
+
+The CDK core framework is responsible for registering and invoking plugins and
+then displaying the formatted validation report. The responsibility of the
+plugin is to act as the translation layer between the CDK framework and the
+policy validation tool. Responsibilities of the plugin may include things like:
+
+- Bundling or installing the policy tool (cfn-guard, opa, etc)
+- Bundling or fetching the rules to evaluate. 
+- Invoking the policy tool, parsing the output, and returning a
+  `ValidationReport` to the framework
+- Handling exemptions
+- Providing contextual information to the policy tool
+
+A plugin can be created in any language supported by CDK. If you are creating a
+plugin that might be consumed by multiple languages then it is recommended
+that you create the plugin in `TypeScript` so that you can use JSII to publish
+the plugin in each CDK language.
+
+#### Developing Plugins
+
+If you need to develop your own policy validation plugin, either because one
+does not exist for your policy tool or because and existing plugin does not meet
+your use case, you start by creating a class that implements the
+`IValidationPlugin` interface from `aws-cdk-lib`.
+
+```ts
+import { ValidationPluginReport } from './report';
+
+/**
+ * Represents a validation plugin that will be executed during synthesis
+ */
+export interface IValidationPlugin {
+  /**
+   * The name of the plugin that will be displayed in the validation
+   * report
+   */
+  readonly name: string;
+
+  /**
+   * The method that will be called by the CDK framework to perform
+   * validations. This is where the plugin will evaluate the CloudFormation
+   * templates for compliance and report and violations
+   */
+  validate(context: IValidationContext): ValidationPluginReport;
+}
+
+/**
+ * Context available to the validation plugin
+ */
+export interface IValidationContext {
+  /**
+   * The absolute path of all templates to be processed
+   */
+  readonly templatePaths: string[];
+}
+
+```
+
+The `validate` method returns an instance of `ValidationPluginReport`, which
+tells the CDK whether the template is compliant, which violations were found
+(if any), and any metadata about the report. These are the report related 
+interfaces:
```

Comment:
nit: this should be in the appendix or design sections, not the working backwards.